### PR TITLE
228 generate presets

### PR DESCRIPTION
**Pull request by robhil**
https://github.com/UXPin/uxpin-merge-tools/issues/228

### CODE REVIEW

**Review by matthewchigira — COMMENTED** [review]
Hey! I've been having a quick look at this today, and at first glance, this is really great! Appreciate all the effort you've put into writing tests!

I'd like to carry on with some more manual testing before we merge though.

One minor thing that I did notice was that you seem to be generating the `import` statement in the preset file from the name of the component marked `default export`, rather than the file name (which is available in the `uxpin.config.js` file)?

There is nothing wrong this assumption you have made, it might be the best way to do it :thinking: But I did notice that if you run this generate-presets on the boilerplate Camera doesn't work because the file is named `Camera.js` but the component is called `CameraPreview`, so it generates an incorrect `import` statement. It's no big deal, but just wanted to flag it up for us to consider...

**Comment by robhil** [comment]
@matthewchigira I will check that issue with importing CameraPreview

**Review by naohiro-t — COMMENTED** [review]
Looking great! Just small comments :)

File: packages/uxpin-merge-cli/src/utils/fs/PresetFileGenerator/helpers/getPropTypeValue.ts [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+import { PropertyType } from '../../../../steps/serialization/component/implementation/ComponentPropertyDefinition';
+import { getDefaultValueForPropType } from './getDefaultValueForPropType';
+
+export function getPropTypeValue(propertyType?:PropertyType):any {
+  if (!propertyType) {
+    return '';
+  }
+
+  switch (propertyType.name) {
+    case 'union':
```
[the remark this answers is not in the document]
Comment (robhil) [reply packages/uxpin-merge-cli/src/utils/fs/PresetFileGenerator/helpers/getPropTypeValue.ts:11]: good catch!

[the remark this answers is not in the document]
Comment (robhil) [reply packages/uxpin-merge-cli/src/utils/fs/PresetFileGenerator/helpers/getPropTypeValue.ts:11]: Hmm I checked PropTypes.oneOf and it returns 'union' 🤔 
Merge team didn't add enum to PropertyTypeStructureMap  as well
https://github.com/UXPin/uxpin-merge-tools/blob/master/packages/uxpin-merge-cli/src/steps/serialization/component/implementation/ComponentPropertyDefinition.ts#L60

[the remark this answers is not in the document]
Comment (naohiro-t) [reply packages/uxpin-merge-cli/src/utils/fs/PresetFileGenerator/helpers/getPropTypeValue.ts]: hmm, strange 🤔 I ran following command with Button from boilerplate which uses `oneOf`

```
naohiro@Naohiros-MacBook-Pro-2 uxpin-merge-boilerplate % ./node_modules/@uxpin/merge-cli/node_modules/.bin/react-docgen src/components/Button/Button.js --pretty
```

```json
{
  "description": "",
  "displayName": "Button",
  "methods": [],
  "props": {
    "onClick": {
      "type": {
        "name": "func"
      },
      "required": false,
      "description": ""
    },
    "disabled": {
      "type": {
        "name": "bool"
      },
      "required": false,
      "description": "",
      "defaultValue": {
        "value": "false",
        "computed": false
      }
    },
    "type": {
      "type": {
        "name": "enum",
        "value": [
          {
            "value": "'primary'",
            "computed": false
          },
          {
            "value": "'secondary'",
            "computed": false
          },
          {
            "value": "'success'",
            "computed": false
          },
          {
            "value": "'error'",
            "computed": false
          },
          {
            "value": "'warning'",
            "computed": false
          }
        ]
      },
      "required": false,
      "description": "",
      "defaultValue": {
        "value": "'primary'",
        "computed": false
      }
    },
    "mode": {
      "type": {
        "name": "enum",
        "value": [
          {
            "value": "'filled'",
            "computed": false
          },
          {
            "value": "'ghost'",
            "computed": false
          },
          {
            "value": "'minimal'",
            "computed": false
          },
          {
            "value": "'flat'",
            "computed": false
          }
        ]
      },
      "required": false,
      "description": "",
      "defaultValue": {
        "value": "'filled'",
        "computed": false
      }
    },
    "title": {
      "type": {
        "name": "string"
      },
      "required": false,
      "description": "@uxpinignoreprop"
    },
    "background": {
      "type": {
        "name": "string"
      },
      "required": false,
      "description": "@uxpinignoreprop"
    },
    "children": {
      "type": {
        "name": "string"
      },
      "required": false,
      "description": "@uxpinpropname Label"
    },
    "icon": {
      "type": {
        "name": "node"
      },
      "required": false,
      "description": ""
    },
    "iconDirection": {
      "type": {
        "name": "enum",
        "value": [
          {
            "value": "'left'",
            "computed": false
          },
          {
            "value": "'right'",
            "computed": false
          }
        ]
      },
      "required": false,
      "description": "",
      "defaultValue": {
        "value": "'left'",
        "computed": false
      }
    },
    "size": {
      "type": {
        "name": "enum",
        "value": [
          {
            "value": "'xs'",
            "computed": false
          },
          {
            "value": "'s'",
            "computed": false
          },
          {
            "value": "'m'",
            "computed": false
          },
          {
            "value": "'l'",
            "computed": false
          },
          {
            "value": "'xl'",
            "computed": false
          },
          {
            "value": "'xxl'",
            "computed": false
          },
          {
            "value": "'xxxl'",
            "computed": false
          }
        ]
      },
      "required": false,
      "description": "",
      "defaultValue": {
        "value": "'m'",
        "computed": false
      }
    },
    "stretched": {
      "type": {
        "name": "bool"
      },
      "required": false,
      "description": "",
      "defaultValue": {
        "value": "true",
        "computed": false
      }
    }
  }
}
```

And TrendLine which uses `oneOfType`
```
naohiro@Naohiros-MacBook-Pro-2 uxpin-merge-boilerplate % ./node_modules/@uxpin/merge-cli/node_modules/.bin/react-docgen src/components/TrendLine/TrendLine.js --pretty
```

```json
Browserslist: caniuse-lite is outdated. Please run next command `yarn upgrade`
{
  "description": "",
  "displayName": "TrendLine",
  "methods": [
    {
      "name": "getDelegatedProps",
      "docblock": null,
      "modifiers": [],
      "params": [],
      "returns": null
    },
    {
      "name": "renderGradientDefinition",
      "docblock": null,
      "modifiers": [],
      "params": [],
      "returns": null
    }
  ],
  "props": {
    "data": {
      "type": {
        "name": "arrayOf",
        "value": {
          "name": "union",
          "value": [
            {
              "name": "number"
            },
            {
              "name": "shape",
              "value": {
                "value": {
                  "name": "number",
                  "required": false
                }
              }
            }
          ]
        }
      },
      "required": true,
      "description": ""
    },
    "smooth": {
      "type": {
        "name": "bool"
      },
      "required": false,
      "description": "",
      "defaultValue": {
        "value": "true",
        "computed": false
      }
    },
    "autoDraw": {
      "type": {
        "name": "bool"
      },
      "required": false,
      "description": "",
      "defaultValue": {
        "value": "true",
        "computed": false
      }
    },
    "autoDrawDuration": {
      "type": {
        "name": "number"
      },
      "required": false,
      "description": "",
      "defaultValue": {
        "value": "1600",
        "computed": false
      }
    },
    "autoDrawEasing": {
      "type": {
        "name": "union",
        "value": [
          {
            "name": "custom",
            "raw": "'ease-in'"
          },
          {
            "name": "custom",
            "raw": "'ease-out'"
          },
          {
            "name": "custom",
            "raw": "'linear'"
          }
        ]
      },
      "required": false,
      "description": "",
      "defaultValue": {
        "value": "'ease-out'",
        "computed": false
      }
    },
    "width": {
      "type": {
        "name": "number"
      },
      "required": false,
      "description": "@uxpinignoreprop"
    },
    "height": {
      "type": {
        "name": "number"
      },
      "required": false,
      "description": "@uxpinignoreprop"
    },
    "padding": {
      "type": {
        "name": "number"
      },
      "required": false,
      "description": "",
      "defaultValue": {
        "value": "8",
        "computed": false
      }
    },
    "radius": {
      "type": {
        "name": "number"
      },
      "required": false,
      "description": "",
      "defaultValue": {
        "value": "10",
        "computed": false
      }
    },
    "gradient": {
      "type": {
        "name": "arrayOf",
        "value": {
          "name": "string"
        }
      },
      "required": false,
      "description": "",
      "defaultValue": {
        "value": "['#00c6ff', '#F0F', '#FF0']",
        "computed": false
      }
    },
    "strokeWidth": {
      "type": {
        "name": "number"
      },
      "required": false,
      "description": "",
      "defaultValue": {
        "value": "3",
        "computed": false
      }
    }
  }
}
```

File: packages/uxpin-merge-cli/src/utils/fs/PresetFileGenerator/PresetFileGenerator.ts [hunk 2]
```diff
@@ -113,10 +113,24 @@ export class PresetFileGenerator {
...
+      return false;
+    }
+  }
+
+  private async generateImports():Promise<string> {
+    const relativeComponentPath:string = relative(this.presetDirectory, this.componentPath)
+      .split('.')
+      .slice(0, -1)
+      .join('.');
+    const importName:string = (await this.hasDefaultExport()) ? this.componentName : `{ ${this.componentName} }`;
```
Comment (matthewchigira) [inline packages/uxpin-merge-cli/src/utils/fs/PresetFileGenerator/PresetFileGenerator.ts:130]: @robhil 

Hmm, looks like that `hasDefaultExport()` doesn't catch components which export like this: `export { Button as default };`. For example the [Button component](https://github.com/uxpin-merge/uxpin-merge-boilerplate/blob/master/src/components/Button/Button.js) of the boilerplate. In this case, the generated preset file now looks like this: `import { Button } from '../Button'`, which causes an error when trying to launch experimental mode:

```
ERROR: Error: Unknown component!
    at Object../src/components/Button/presets/0-default.jsx (/home/matthew/Documents/UXPin/Development/test/boilerplate/uxpin-merge-boilerplate/.uxpin-merge/webpack:/src/co
mponents/Button/presets/0-default.jsx:5:3)
    at __webpack_require__ (/home/matthew/Documents/UXPin/Development/test/boilerplate/uxpin-merge-boilerplate/.uxpin-merge/webpack:/webpack/bootstrap:19:1)
    at Object../.uxpin-merge/presets-70e9b712-e569-41c9-b377-bb7962c6b94f.js (/home/matthew/Documents/UXPin/Development/test/boilerplate/uxpin-merge-boilerplate/.uxpin-merg
e/__bundle__presets-70e9b712-e569-41c9-b377-bb7962c6b94f.js:197:102)
    at __webpack_require__ (/home/matthew/Documents/UXPin/Development/test/boilerplate/uxpin-merge-boilerplate/.uxpin-merge/webpack:/webpack/bootstrap:19:1)
```

Manually removing the curly braces fixes the error.

File: packages/uxpin-merge-cli/src/utils/fs/PresetFileGenerator/PresetFileGenerator.ts [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+    }
+
+    if (await pathExists(this.presetFilePath)) {
+      printWarning(`File ${this.presetFilePath} exists`);
+      return;
+    }
+
+    const componentFileContent:string = await this.generateComponentFile();
+    await writeToFile(this.presetFilePath, componentFileContent);
+    printLine(`File ${this.presetFilePath} created successfully`, { color: PrintColor.GREEN });
```
Comment (matthewchigira) [inline packages/uxpin-merge-cli/src/utils/fs/PresetFileGenerator/PresetFileGenerator.ts:76]: @robhil 

Another very minor thing, but for consistency with other messages that the CLI prints, could you utilise the green tick box ✅ for successful messages like this one, and the warning triangle 🛑 or info sign 🛈 for messages that say that the file already exists. You can search the project for examples. Thanks!

File: packages/uxpin-merge-cli/src/utils/fs/PresetFileGenerator/helpers/getPropTypeValue.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (robhil) [reply packages/uxpin-merge-cli/src/utils/fs/PresetFileGenerator/helpers/getPropTypeValue.ts]: yeah I see it returns enum but when I call generate presets it's working fine for Button and TrendLine so I would not change this for now :)

File: packages/uxpin-merge-cli/src/utils/fs/PresetFileGenerator/PresetFileGenerator.ts (hunk 3 above)
Comment (robhil) [reply packages/uxpin-merge-cli/src/utils/fs/PresetFileGenerator/PresetFileGenerator.ts:76]: I saw that for warnings we used 👉 so I used it as well

**Review by naohiro-t — APPROVED** [review]
Looks good to me 🚀

**Review by matthewchigira — APPROVED** [review]
Looks good to me :rocket:

**Pull request merged**